### PR TITLE
Fix off-by-one warning from LGTM and add tests for NodePort range

### PR DESCRIPTION
**Pull request by christarazi**
Fixes warning as seen here: https://lgtm.com/projects/g/cilium/cilium/snapshot/5770266093fca7b394682a3a2ba6a21f92ec07da/files/pkg/option/config.go?sort=name&dir=ASC&mode=heatmap#xd94e5a86fe539522:1

https://github.com/cilium/cilium/blob/dd534bc55947d54b8071ae067238ae9cdbe1b66b/pkg/option/config.go#L1654

Side note: I was looking to add test coverage relevant to the change and noticed that [DaemonConfig.Populate()](https://github.com/cilium/cilium/blob/dd534bc55947d54b8071ae067238ae9cdbe1b66b/pkg/option/config.go#L1654) is not tested. After investigating a bit, I came to the conclusion that I'll cover it in a separate PR. This is because `Populate()` does not return an error and instead does `log.Fatal` when something goes wrong. This made writing unit tests to cover non-happy path more difficult. If refactoring `Populate()` to return an `error` seems reasonable, I will go ahead with that.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10151)

### CODE REVIEW

**Review by aanm — APPROVED** [review]
Thanks for the PR @christarazi !

**Comment by aanm** [comment]
test-me-please

**Comment by christarazi** [comment]
Ah, I see NodePortRange is not a required parameter. There's an assertion that failed outside the `option` package. Fixing now and will re-run all unit tests.

**Comment by christarazi** [comment]
Squashed into one commit.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by christarazi** [comment]
Looks like it failed again. Send it off again. I'm assuming this doesn't have to do with this PR, correct?

**Comment by christarazi** [comment]
test-me-please

**Comment by joestringer** [comment]
Looks like the failures are quite interesting:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17081/testReport/junit/Suite-k8s-1/11/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_flags_send_notifications/
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17081/testReport/junit/Suite-k8s-1/11/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_restricts_notifications/

Specifically:
```
Top 5 errors/warnings:
Error while initializing daemon
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/cilium/bpf/lib/utils.h:124:45: note: expanded from macro 'build_bug_on'
# define build_bug_on(e) ((void)sizeof(char[1 - 2*!!(e)]))
                                            ^~~~~~~~~~~
Cilium pods: [cilium-2wqlm cilium-b49fj]
```

I think it's worth digging deeper into the attached zip for one of these, and looking at the specific build failure to determine whether it's nodeport related. It's plausible that this userspace change has affected the way that we generate the `#define`s in the BPF, leading to this build-time compile failure on the k8s-1.11 (newer kernel / kube-proxy-free) test runs.

**Comment by christarazi** [comment]
Ah, thanks for the pointer. Looking into it now...

**Comment by christarazi** [comment]
Looking through the [first Jenkins build link from above](https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17081/testReport/junit/Suite-k8s-1/11/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_flags_send_notifications/) inside `cilium-test.log`, I see:

```
level=warning msg="/var/lib/cilium/bpf/lib/nodeport.h:1269:3: error: array size is negative" subsys=datapath-loader
level=warning msg="build_bug_on(!(NODEPORT_PORT_MIN     < NODEPORT_PORT_MAX));" subsys=datapath-loader
```

I wanted to understand what that meant, and learned `build_bug_on` is a macro that tries to create a compile-time error to validate the bpf program before loading:

```c
#define build_bug_on(e) ((void)sizeof(char[1 - 2*!!(e)]))

int main()
{
    build_bug_on(!(1 < 2)); // OK
    build_bug_on(!(3 < 2)); // Error
    build_bug_on(!(2 < 2)); // Error
}
```

The lines marked as errors trigger the exact error message as seen in the logs. This tells me that somehow `NODEPORT_PORT_MIN` was parsed out to be less than or equal to `NODEPORT_PORT_MAX`. It's also strange given that NodePortRange has defaults.

Not sure how this can happen given the code, but I will continue investigating by trying to re-run the E2E tests locally in Vagrant. Any tips for quicker feedback?

**Comment by joestringer** [comment]
I'm actually a little surprised how few pieces of information are present in the artifacts from these environments, usually I would suggest digging into the `/var/run/cilium/state` copy to see what the generated headerfiles look like.

I wonder if you would be able to deploy either minikube with a new enough kernel or use something like microk8s to set up Cilium in a recent (Linux 4.19+) environment and observe the same behaviour. I suspect that the node port piece here would be enabled by default in such cases so simply deploying Cilium with this PR's code should trigger it. This test must be configuring something slightly differently though so you'd want to ensure that those settings are matching your local environment. Looking at the [test code](https://github.com/cilium/cilium/blob/f5c42e7e0cda0c2a0ee46a8467382c8acc6fd388/test/k8sT/DatapathConfiguration.go#L137) should help with this.

Of course, the Vagrant E2E tests is basically doing this too (Provided that you ensure that the `NET_NEXT` variable is true so that the newer kernel image is pulled) so maybe that's simpler. For sure you would want to avoid running the full e2e tests if possible and just focus these tests that are failing.

Taking a step back, this PR assumes that `viper.IsSet(NodePortRange)` is equivalent to `len(nodePortRange) > 0` but I wonder if due to either viper environment variable binding (`BindEnv()`) or the way that the configmap configures the values for these options, the actual execution is different from what you expected here. My guess is that somewhere in `pkg/datapath/linux/config`, these values are not being written correctly any more. While you're waiting for the above setup to spin up you could either (1) try to eyeball that code and understand what's different, or (2) see if there's a way you could wrap some unit-testing around this specific code that's being changed to evaluate the difference in inputs/outputs.

**Comment by christarazi** [comment]
Thanks for the thoroughness of your response. 

I was actually suspecting `viper.IsSet` as well. It's the only potential path I see so far. I think what could be happening is `NodePortMin` and `NodePortMax` are both set to their Golang zero values (which would simply be `0`). That would definitely trigger the C macro to force a compile-time error.

I will try for deploying Cilium directly, rather than E2E tests as those seem a bit too heavy.

**Comment by christarazi** [comment]
I figured out what's happening here. It was due to `viper.IsSet` returning false. I was able to confirm by doing some delve debugging.

I reverted the original change, separated out the NodePortRange logic from `Populate()` so that it can be more easily unit tested, wrote unit tests to cover the code, and brought back what should be the correct logic to solve the original issue.

For posterity, here's explanation of why `viper.IsSet` caused the observed behavior:

Essentially, `viper.IsSet` searches if a key exists in its `override` map. This map is only written to if you call `viper.Set`. If the key does not exist in the `override` map, then `viper.IsSet` returns false. In other words, `viper.IsSet` informs you if the user has explicitly passed in the parameter (via CLI, env-var, etc). The gotcha here was that defaulted parameters will have `viper.IsSet` == `false`.

This makes sense and is expected behavior. We technically didn't "set" the NodePortRange explicitly. In the end, the assumption that having defaults makes `viper.IsSet` return true was incorrect.

**Comment by christarazi** [comment]
For efficiency, in commit https://github.com/cilium/cilium/pull/10151/commits/8cf8b80f8c60d143d903cca91e96f93533e52bf3, I left some TODOs so that I can ask gather feedback from the team, while allowing CI to catch any potential problems in the meanwhile.

Onto the actual question, as https://github.com/cilium/cilium/pull/10151/commits/8cf8b80f8c60d143d903cca91e96f93533e52bf3 states in the TODOs, this [test](https://github.com/cilium/cilium/blob/640b7a6cb35905ca1aafd3010baf9b8265d383c9/pkg/envoy/envoy_test.go#L68) requires that NodePortRange to be an empty slice. Also, the code before this original change did not assert to see if a user passed in an empty slice.

Is this intended behavior? If I'm not mistaken, the consequences of this is that a user can explicitly pass NodePortRange as an empty slice and the code will happily go on, without warning and the defaults are overridden.

**Comment by joestringer** [comment]
test-me-please

File: pkg/option/config.go
Comment (joestringer) [inline pkg/option/config.go]: I'm a bit surprised that this is the case. /cc @jrajahalme do you know why `pkg/envoy_test.go` would rely on this?

File: pkg/option/config.go
Comment (joestringer) [inline pkg/option/config.go]: nit: this `if` condition can be folded into the `else` check above.

**Review by joestringer — CHANGES_REQUESTED** [review]
CI is happy now :tada: 

The added unit tests make this much easier to look at and validate, thanks for adding those. The separate refactoring commit also made review easier.

I think you can drop the first two commits, and with that the commit messages in the later commits don't really make sense given that the code doesn't use `viper.IsSet()`.  I found the explanation useful specifically for reviewing this PR, but it's probably not necessary to leave in the codebase just given how infrequently we need to deal with this detail of the library.

I don't have an answer for you on the expectations from the envoy_test, but given your explanation that the defaults we configure will set the array's values correctly then it's probably valid for the final patch's functional changes to be omitted? Feel free to ping me on Slack tomorrow if you want a tighter turn-around with the CI.

File: pkg/option/config.go
Comment (christarazi) [reply pkg/option/config.go]: Fixed: cd702be

File: pkg/option/config.go
Comment (christarazi) [reply pkg/option/config.go]: After taking another look, I'm going to take a stab at explaining. 

envoy_test.go uses a "preconfigured" `option.Config`: https://github.com/cilium/cilium/blob/640b7a6cb35905ca1aafd3010baf9b8265d383c9/pkg/envoy/envoy_test.go#L68

Then it calls `Populate` which will set additional values in `Config` with values from viper. Due to envoy_test.go running in the test context, the defaults *for* viper haven't been set. They haven't been set because the code which does set defaults is in `daemon_main.go`'s `init()`:

https://github.com/cilium/cilium/blob/5b9002d61500b674c04b5bdd4aa00f91619af8e0/daemon/daemon_main.go#L187

In a non-testing context (running normally), the `init()` mentioned above would run before `Populate` gets called, and everything is OK.

Therefore, in the testing context `Populate` runs standalone meaning that the `init()` mentioned above does NOT get executed. This results in three things: there are no defaults, it has nothing to retrieve from viper, and NodePortRange stays empty.

This is why you can see in the unit tests I wrote (d8f9160), all the pre-test setups are setting up defaults to inform viper, simply because the input to the function under test is viper itself.  

The only reason envoy_test.go complained is because the unit tests I wrote checked for an empty NodePortRange. The code before this PR silently allowed that, whether by accident or on purpose. Either way, it is quite tricky to wrap your head around, but I hope that cleared it up. 

It feels wrong to allow NodePortRange to be empty. I think this all just may have been a coincidental result. We can modify envoy_test.go to reduce its dependency on the entire `DaemonConfig` struct, and only just the options it needs. However I recognize that we may be getting into yak shaving territory here. Let me know what you think.

**Comment by christarazi** [comment]
> I think you can drop the first two commits, and with that the commit messages in the later commits don't really make sense given that the code doesn't use `viper.IsSet()`. I found the explanation useful specifically for reviewing this PR, but it's probably not necessary to leave in the codebase just given how infrequently we need to deal with this detail of the library.

Great, that was the intent. Agreed on rewording the commits.
 
> it's probably valid for the final patch's functional changes to be omitted? 

I don't fully follow what you meant here.

File: pkg/option/config.go
Comment (christarazi) [reply pkg/option/config.go:1970]: This error string is directly used as the log message in the caller. I can change this if people would like.

File: pkg/option/config.go
Comment (joestringer) [reply pkg/option/config.go]: Oh, yeah this seems a bit broken for `pkg/envoy` unit-testing to be relying upon CLI args, but I'm not all that into yak-shaving so it's probably fine to leave it as-is.

There's no big danger here that the guard rails arounnd CLI validation cause inexperienced users to waste lots of time chasing down issues with this, the nodeport range is a pretty advanced configuration and anyone configuring it should likely have an idea about the correct way to pass the values in.

**Comment by joestringer** [comment]
> > it's probably valid for the final patch's functional changes to be omitted?

> I don't fully follow what you meant here.

Oh, I was musing whether we need the change to only return an error if the number of args is > 0. It sounds like the default settings will always set 2 args, so then it's a question whether we should allow users to specify the option but leave it empty. I don't think it ultimately makes a big difference so we can probably just drop the TODO. Otherwise LGTM.

**Comment by joestringer** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Looks like it hit known CI flake #9297 , this is not the same as the failure that we were seeing earlier in this PR development. LGTM, will just wait until we open the merge window for v1.8 development PRs.

**Review by aanm — COMMENTED** [review]
@christarazi it seems we are introducing the usage of `errors.Errorf` with this change. Is there a reason for this? cc @joestringer

**Comment by christarazi** [comment]
> @christarazi it seems we are introducing the usage of `errors.Errorf` with this change. Is there a reason for this? cc @joestringer

No specific reason, I think it was just an oversight. I will change to `fmt.Errorf` instead.

**Comment by joestringer** [comment]
test-me-please

**Pull request merged**
